src/state.rs

```diff
@@ -109,6 +113,54 @@ pub struct ResolvedDns {
     dns_refresh_rate: std::time::Duration,
 }
 
+#[derive(serde::Serialize, Default, Debug)]
+pub struct SingleFlight {
+    // is_first_req_in can record requests number and it can be used to
+    // check the request number, such as the first request.
+    is_first_req_in: bool,
+    // mutex and condvar can implement the multiple waiters and a single notifier
+    #[serde(skip_serializing)]
+    mutex: Arc<Mutex<()>>,
+    #[serde(skip_serializing)]
+    condvar: Arc<Condvar>,
+}
+
+impl Clone for SingleFlight {
+    fn clone(&self) -> Self {
+        SingleFlight {
+            is_first_req_in: self.is_first_req_in,
+            mutex: self.mutex.clone(),
+            condvar: self.condvar.clone(),
+        }
+    }
+}
+
+impl SingleFlight {
+    pub fn new() -> Self {
+        SingleFlight {
+            // is_first_req_in is set to true because
+            // there always is the first request
+            is_first_req_in: true,
+            mutex: Arc::new(Mutex::new(())),
+            condvar: Arc::new(Condvar::new()),
+        }
+    }
+
+    pub fn wait_for_notifying(&self) {
+        let timeout_millis = std::time::Duration::from_millis(100);
+        let lock = self.mutex.lock().unwrap();
+
+        // Must define _unused to avoid immediately dropping of the lock and use wait_timeout with a timeout of 100 milliseconds
+        let (_is_started, _timeout_result) =
+            self.condvar.wait_timeout(lock, timeout_millis).unwrap();
```

Comment:
is there a similar condvar in tokio

Reply:
Hi @nmittler [here ](https://github.com/istio/ztunnel/pull/672#discussion_r1355979606) is the exact place to record why we need to use `wait_timeout`. Hope this can be helpful.

Reply:
@zhlsunshine I'm not sure that thread addresses my general concern of blocking a tokio thread for this. Using a timeout still means that a single invocation is blocking the thread for up to 100 milliseconds, which is a very long time. If multiple requests come in at the same time, then we could effectively take down the tokio runtime. I suspect we should be using tokio synchronization primitives for this.

Reply:
Yeah this is broken; it will block the entire thread. We only have 2 threads. Async code shouldn't block

Reply:
Thanks @nmittler & @howardjohn for your comments and information sharing!  Yeah, it's fairly coarse by using `std::sync::Mutex`, and `tokio::sync::Notify` may result in spurious wakeups for abnormal cases especially for unexpected pending notification. How about to use `tokio::sync::Notify`, and remove the `Notify` instance from `sf_by_hostname` once current DNS resolving task completed. Moreover, the waiters will register the waiting list before the notifier because of the DNS resolving task.  Please check the later PR.

Reply:
Thanks @nmittler & @howardjohn for your comments and information sharing!  Yeah, it's fairly coarse by using `std::sync::Mutex`, and `tokio::sync::Notify` may result in spurious wakeups for abnormal cases especially for unexpected pending notification. How about to use `tokio::sync::Notify`, and remove the `Notify` instance from `sf_by_hostname` once current DNS resolving task completed. Moreover, the waiters will register the waiting list before the notifier because of the DNS resolving task.  Please help to review the later PR.